File: ios/Runner/CryptoChannelHandler.swift
//
//  CryptoChannelHandler.swift
//  Runner
//
//  Created by Oluwakemi Mafe on 2026-08-16.
//

import Flutter
import Foundation

final class CryptoChannelHandler {
    private let cryptoService: CryptoServicing
    
    init( cryptoService: CryptoServicing = CryptoService()) {
        self.cryptoService = cryptoService
    }

    func handle(
        call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        do {
            switch call.method {
            case "encrypt":
                try encrypt(call: call, result: result)

            case "decrypt":
                try decrypt(call: call, result: result)

            case "deleteKey":
                try cryptoService.deleteKey()
                result(nil)

            default:
                result(FlutterMethodNotImplemented)
            }
        } catch CryptoServiceError.authenticationRequired {
            result(
                FlutterError(
                    code: "AUTHENTICATION_REQUIRED",
                    message: "User authentication is required to use the vault encryption key.",
                    details: nil
                )
            )
        } catch {
            result(
                FlutterError(
                    code: "CRYPTO_ERROR",
                    message: error.localizedDescription,
                    details: nil
                )
            )
        }
    }

    private func encrypt(
        call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) throws {
        guard
            let arguments = call.arguments as? [String: Any],
            let plainText = arguments["plainText"] as? String
        else {
            throw ChannelError.invalidArguments
        }

        result(
            try cryptoService.encrypt(plainText)
        )
    }

    private func decrypt(
        call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) throws {
        guard
            let arguments = call.arguments as? [String: Any],
            let cipherText = arguments["cipherText"] as? [UInt8],
            let nonce = arguments["nonce"] as? [UInt8],
            let mac = arguments["mac"] as? [UInt8]
        else {
            throw ChannelError.invalidArguments
        }

        result(
            try cryptoService.decrypt(
                cipherText: cipherText,
                nonce: nonce,
                mac: mac
            )
        )
    }
}

enum ChannelError: Error {
    case invalidArguments
}

File: android/app/src/main/kotlin/com/example/key_vault/CryptoChannelHandler.kt
package com.oluwakemimafe.key_vault

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class CryptoChannelHandler(
    private val cryptoService: CryptoService =
        CryptoService()
) {

    fun handle(
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        try {
            when (call.method) {
                "unlockSession" -> {
                    cryptoService.unlockSession()
                    result.success(null)
                }

                "lockSession" -> {
                    cryptoService.lockSession()
                    result.success(null)
                }

                "encrypt" -> {
                    encrypt(
                        call,
                        result
                    )
                }

                "decrypt" -> {
                    decrypt(
                        call,
                        result
                    )
                }

                "deleteKey" -> {
                    cryptoService.deleteKey()
                    result.success(null)
                }

                else -> {
                    result.notImplemented()
                }
            }
        } catch (
            exception: AuthenticationRequiredException
        ) {
            result.error(
                "AUTHENTICATION_REQUIRED",
                exception.message,
                null
            )
        } catch (exception: Exception) {
            result.error(
                "CRYPTO_ERROR",
                exception.message
                    ?: "Native cryptographic operation failed.",
                null
            )
        }
    }

    private fun encrypt(
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        val plainText =
            call.argument<String>(
                "plainText"
            )
                ?: throw IllegalArgumentException(
                    "plainText is required."
                )

        result.success(
            cryptoService.encrypt(
                plainText
            )
        )
    }

    private fun decrypt(
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        val cipherText =
            call.argument<List<Int>>(
                "cipherText"
            )
                ?: throw IllegalArgumentException(
                    "cipherText is required."
                )

        val nonce =
            call.argument<List<Int>>(
                "nonce"
            )
                ?: throw IllegalArgumentException(
                    "nonce is required."
                )

        val mac =
            call.argument<List<Int>>(
                "mac"
            )
                ?: throw IllegalArgumentException(
                    "mac is required."
                )

        result.success(
            cryptoService.decrypt(
                cipherText.toByteArray(),
                nonce.toByteArray(),
                mac.toByteArray()
            )
        )
    }

    private fun List<Int>.toByteArray(): ByteArray =
        ByteArray(size) { index ->
            this[index].toByte()
        }
}